stetho/src/main/java/com/facebook/stetho/inspector/elements/UIDOMProvider.java

```diff
@@ -0,0 +1,101 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
+
+package com.facebook.stetho.inspector.elements;
+
+import android.app.Activity;
+import android.app.Application;
+import android.content.Context;
+import android.view.View;
+import android.view.ViewGroup;
+import android.view.Window;
+import android.widget.TextView;
+
+import com.facebook.stetho.inspector.protocol.module.DOM;
+
+public final class UIDOMProvider implements DOM.Provider {
+  private final NodeDescriptorMap mNodeDescriptorMap;
+  private final Application mApplication;
+  private Listener mListener;
+
+  public UIDOMProvider(Context context) {
+    mApplication = (Application)context.getApplicationContext();
+
+    mNodeDescriptorMap = new NodeDescriptorMap()
+        .beginInit()
+        .register(Activity.class, new ActivityNodeDescriptor())
+        .register(Application.class, new ApplicationNodeDescriptor())
+        .register(Object.class, new ObjectNodeDescriptor())
+        .register(TextView.class, new TextViewNodeDescriptor())
+        .register(View.class, new ViewNodeDescriptor())
+        .register(ViewGroup.class, new ViewGroupNodeDescriptor())
+        .register(Window.class, new WindowNodeDescriptor())
+        .setListener(new NodeDescriptorListener())
+        .endInit();
+  }
+
+  // Forwards notifications from NodeDescriptor to DOM
+  private class NodeDescriptorListener implements NodeDescriptor.Listener {
+    @Override
+    public void onAttributeModified(Object element, String name, String value) {
+      mListener.onAttributeModified(element, name, value);
+    }
+
+    @Override
+    public void onAttributeRemoved(Object element, String name) {
+      mListener.onAttributeRemoved(element, name);
+    }
+
+    @Override
+    public void onChildInserted(Object parentElement, Object previousElement, Object childElement) {
+      mListener.onChildInserted(parentElement, previousElement, childElement);
+    }
+
+    @Override
+    public void onChildRemoved(Object parentElement, Object childElement) {
+      mListener.onChildRemoved(parentElement, childElement);
+    }
+  }
+
+  @Override
+  public void setListener(Listener listener) {
+    mListener = listener;
+  }
+
+  /*private static boolean canEnableHighlight(Activity activity) {
+    Context context = activity.getApplicationContext();
+    if (context == null) {
+      return false;
+    }
+
+    int permission = context.checkCallingOrSelfPermission(Manifest.permission.SYSTEM_ALERT_WINDOW);
+    return permission != PackageManager.PERMISSION_DENIED;
+  }*/
+
+  @Override
+  public Object getRootElement() {
+    return mApplication;
+  }
+
+  @Override
+  public DOM.NodeDescriptor getNodeDescriptor(Object element) {
+    return (element == null) ? null : mNodeDescriptorMap.get(element.getClass());
+  }
+
+  @Override
+  public void highlightElement(Object element) {
+    if (element instanceof View) {
```

Comment:
I scratched my head on the right way to do this for a bit, and finally decided on mirroring the DOM JSON API as much as possible in the Java code; highlightNode and hideHighlight are methods on the JSON DOM protocol, therefore they're methods on our DOM.Provider. This should make it easier to navigate and understand the two if they are mirrors of each other. Plus, from a practical standpoint, this would require the NodeDescriptor to coordinate with the DOM.Provider. There isn't currently a channel of communication in that direction and I wasn't sure this was the right justification for adding it. It's very easy for a DOM.Provider implementation to delegate this call to NodeDescriptor if that's the strategy it chooses to use (just use getNodeDescriptor!), but the reverse isn't as easy (for NodeDescriptor to delegate to DOM.Provider).